sdk/communication/Azure.Communication.Administration/src/Models/InternalPhoneNumberReservationOperation.cs

```diff
@@ -0,0 +1,135 @@
+
+        /// <summary>
+        /// Check for the latest status of the operation.
+        /// </summary>
+        /// <param name="cancellationToken">
+        /// Optional <see cref="CancellationToken"/> to propagate
+        /// notifications that the operation should be cancelled.
+        /// </param>
+        /// <returns>The <see cref="Response"/> with the status update.</returns>
+        internal Response UpdateStatus(CancellationToken cancellationToken = default)
+            => UpdateStatusAsync(false, cancellationToken).EnsureCompleted();
+
+        /// <summary>
+        /// Check for the latest status of the operation.
+        /// </summary>
+        /// <param name="cancellationToken">
+        /// Optional <see cref="CancellationToken"/> to propagate
+        /// notifications that the operation should be cancelled.
+        /// </param>
+        /// <returns>The <see cref="Response"/> with the status update.</returns>
+        internal async ValueTask<Response> UpdateStatusAsync(CancellationToken cancellationToken = default)
+            => await UpdateStatusAsync(true, cancellationToken).ConfigureAwait(false);
+
+        /// <summary>
+        /// Check for the latest status of the operation.
+        /// </summary>
+        /// <param name="cancellationToken">
+        /// Optional <see cref="CancellationToken"/> to propagate
+        /// notifications that the operation should be cancelled.
+        /// </param>
+        /// <param name="async" />
+        /// <returns>The <see cref="Response"/> with the status update.</returns>
+        private async ValueTask<Response> UpdateStatusAsync(bool async, CancellationToken cancellationToken)
+        {
+            if (HasCompleted)
+            {
+                return GetRawResponse();
+            }
+
+            if (cancellationToken == default)
+            {
+                cancellationToken = _cancellationToken;
+            }
+
+            Response<PhoneNumberReservation> update = async
+                ? await _client.GetReservationByIdAsync(reservationId: Id, cancellationToken: cancellationToken).ConfigureAwait(false)
+                : _client.GetReservationById(reservationId: Id, cancellationToken: cancellationToken);
+
+            var response = update.GetRawResponse();
+            _rawResponse = response;
+
+            if (IsOperationFinished(update))
+            {
+                lock (_lockObject)
+                {
+                    _finalRawResponse = response;
+                    _value = update.Value;
+                    _hasCompleted = true;
+                }
+            }
+
+            return _rawResponse;
+        }
+
+        private bool IsOperationFinished(Response<PhoneNumberReservation> response)
+        {
+            if (!response.Value.Status.HasValue)
+                return false;
+
+            return _terminateStatuses.Contains(response.Value.Status.Value);
```

Comment:
I think we need to distinguish between Success and Failure terminate statuses.

If we encounter `Expired`, for example, the LRO shouldn't complete and provide the value. Instead we should throw here with a meaningful exception.

KeyVault itself isn't consistent with all its LROs but here's an example of what I described above
https://github.com/Azure/azure-sdk-for-net/blob/0eb893cfda1a9b39801d414ec77ce753eec43868/sdk/keyvault/Azure.Security.KeyVault.Keys/src/DeleteKeyOperation.cs#L126

@christothes Do I understand the expected failure handling right?

Reply:
Typically we would throw when the `Value` property is evaluated if either the operation has not completed or completed unsuccessfully.